cpp/src/rolling/detail/range_utils.cuh

```diff
@@ -413,209 +413,226 @@ template <typename T, typename V>
+template <window_tag WindowTag,
+          direction Direction,
+          cudf::order Order,
+          typename Grouping,
+          typename OrderbyT,
+          typename DeltaT>
+struct distance_kernel {
+  distance_kernel(Grouping groups,
+                  DeltaT const* row_delta,
+                  column_device_view::const_iterator<OrderbyT> begin,
+                  column_device_view::const_iterator<OrderbyT> end)
+    : groups{groups}, row_delta{row_delta}, begin{begin}, end{end}
+  {
+  }
+  Grouping groups;  ///< Group information to determine bounds on current row's window
+  static_assert(cuda::std::is_same_v<Grouping, ungrouped> ||
+                  cuda::std::is_same_v<Grouping, grouped> ||
+                  cuda::std::is_same_v<Grouping, ungrouped_with_nulls> ||
+                  cuda::std::is_same_v<Grouping, grouped_with_nulls>,
+                "Invalid grouping descriptor");
+  DeltaT const* row_delta;  ///< Delta from current row that defines the interval endpoint. This
+                            ///< pointer is null for UNBOUNDED and CURRENT_ROW windows.
+  column_device_view::const_iterator<OrderbyT> begin;  ///< Iterator to beginning of orderby column
+  column_device_view::const_iterator<OrderbyT> end;    ///< Iterator to end of orderby column
+
   /**
-   * @brief Functor to compute distance from a given row to the edge
-   * of the window.
-   *
-   * @tparam Grouping Object defining how the orderby column is
-   * grouped.
-   * @tparam OrderbyT Type of elements in the orderby columns.
-   * @tparam DeltaT Type of the elements in the scalar delta (returned
-   * by scalar.data()).
-   * @param groups The grouping object.
-   * @param row_delta Pointer to row delta on device.
-   * @param begin Iterator to begin of orderby column on device.
-   * @param end Iterator to end of orderby column on device.
-   *
-   * @note Let `x` be the value of the current row and `delta` the provided
-   * row delta then for bounded windows the endpoints are computed as follows.
-   *
-   *           | ASCENDING | DESCENDING
-   * ----------+-----------+-----------
-   * PRECEDING | x - delta | x + delta
-   * FOLLOWING | x + delta | x - delta
+   * @brief Compute the offset to the end of the window.
    *
-   * See `saturating_sub` and `saturating_add` for details of the implementation of
-   * saturating addition/subtraction.
+   * @param i The current row index.
+   * @return Offset to the current row's window endpoint.
    */
-  template <typename Grouping, typename OrderbyT, typename DeltaT>
-  struct distance_kernel {
-    distance_kernel(Grouping groups,
-                    DeltaT const* row_delta,
-                    column_device_view::const_iterator<OrderbyT> begin,
-                    column_device_view::const_iterator<OrderbyT> end)
-      : groups{groups}, row_delta{row_delta}, begin{begin}, end{end}
-    {
+  template <window_tag W = WindowTag, CUDF_ENABLE_IF(W == window_tag::UNBOUNDED)>
```

Comment:
I found having separate SFINAE overloads for different window types to be more readable despite being more verbose. They almost feel like different algorithms.